Source/Core/VideoBackends/D3D12/D3DUtil.cpp

```diff
@@ -0,0 +1,1063 @@
+			state_after   // UINT StateAfter;
+		}
+	};
+
+	command_list->ResourceBarrier(1, &resourceBarrierDesc);
+}
+
+// Ring buffer class, shared between the draw* functions
+class UtilVertexBuffer
+{
+public:
+	explicit UtilVertexBuffer(int size) : m_max_size(size)
+	{
+		CheckHR(
+			device12->CreateCommittedResource(
+				&CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_UPLOAD),
+				D3D12_HEAP_FLAG_NONE,
+				&CD3DX12_RESOURCE_DESC::Buffer(m_max_size),
+				D3D12_RESOURCE_STATE_GENERIC_READ,
+				nullptr,
+				IID_PPV_ARGS(&m_buf12)
+				)
+			);
+
+		// Map buffer for CPU access upon creation.
+		// On D3D12, CPU-visible resources are persistently mapped.
+
+		CheckHR(m_buf12->Map(0, nullptr, &m_buf12_data));
+	}
+	~UtilVertexBuffer()
+	{
+		D3D::command_list_mgr->DestroyResourceAfterCurrentCommandListExecuted(m_buf12);
+	}
+
+	int GetSize() const { return m_max_size; }
+
+	int GetRoomLeftInBuffer() const { return m_max_size - m_offset; }
+
+	// returns vertex offset to the new data
+	int AppendData(void* data, int size, int vertex_size)
+	{
+		if (m_offset + size >= m_max_size)
+		{
+			// Wrap buffer around.
+			m_offset = 0;
+		}
+
+		m_offset = ((m_offset +vertex_size-1)/vertex_size)*vertex_size; // align offset to vertex_size bytes
+
+		memcpy(static_cast<u8*>(m_buf12_data) + m_offset, data, size);
+
+		m_offset += size;
+		return (m_offset - size) / vertex_size;
+	}
+
+	int BeginAppendData(void** write_ptr, int size, int vertex_size)
+	{
+		_dbg_assert_(VIDEO, size < m_max_size);
+
+		int aligned_offset = ((m_offset + vertex_size - 1) / vertex_size) * vertex_size; // align offset to vertex_size bytes
+		if (aligned_offset + size > m_max_size)
+		{
+			// wrap buffer around and notify observers
+			m_offset = 0;
+			aligned_offset = 0;
+		}
+
+		*write_ptr = reinterpret_cast<byte*>(m_buf12_data) + aligned_offset;
+		m_offset = aligned_offset + size;
+		return aligned_offset / vertex_size;
+	}
+
+	void EndAppendData()
+	{
+		// No-op on DX12.
+	}
+
+	ID3D12Resource* GetBuffer12()
+	{
+		return m_buf12;
+	}
+
+private:
+	ID3D12Resource* m_buf12 = nullptr;
+	void* m_buf12_data = nullptr;
+
+	int m_offset = 0;
+	int m_max_size = 0;
+};
+
+CD3DFont font;
+UtilVertexBuffer* util_vbuf_stq = nullptr;
+UtilVertexBuffer* util_vbuf_cq = nullptr;
+UtilVertexBuffer* util_vbuf_clearq = nullptr;
+UtilVertexBuffer* util_vbuf_efbpokequads = nullptr;
+
+static const unsigned int s_max_num_vertices = 8000 * 6;
+
+struct FONT2DVERTEX
+{
+	float x, y, z;
+	float col[4];
+	float tu, tv;
+};
+
+FONT2DVERTEX InitFont2DVertex(float x, float y, u32 color, float tu, float tv)
+{
+	FONT2DVERTEX v;   v.x=x; v.y=y; v.z=0;  v.tu = tu; v.tv = tv;
+	v.col[0] = ((float)((color >> 16) & 0xFF)) / 255.f;
+	v.col[1] = ((float)((color >>  8) & 0xFF)) / 255.f;
+	v.col[2] = ((float)((color >>  0) & 0xFF)) / 255.f;
+	v.col[3] = ((float)((color >> 24) & 0xFF)) / 255.f;
+	return v;
+}
+
+CD3DFont::CD3DFont() : m_tex_width(512), m_tex_height(512)
+{
+	m_texture12 = nullptr;
```

Comment:
all of these initializations can be done directly on the class members in the class body.